Build failure: awscli2

### Steps To Reproduce

Steps to reproduce the behavior:
1. build *awscli2*

### Build log

```
Running phase: configurePhase
patching script interpreter paths in ./configure
./configure: interpreter directive changed from "#! /bin/sh" to "/nix/store/5lr5n3qa4day8l1ivbwlcby2nknczqkq-bash-5.2p26/bin/sh"
configure flags: --prefix=/nix/store/gh4r6pp80ghgznzm57q096c3k9rckwq6-awscli2-2.15.15 --bindir=/nix/store/gh4r6pp80ghgznzm57q096c3k9rckwq6-awscli2-2.15.15/bin --sbindir=/nix/store/gh4r6pp80ghgznzm57q096c3k9rckwq6-awscli2-2.15.15/sbin --includedir=/nix/store/gh4r6pp80ghgznzm57q096c3k9rckwq6-awscli2-2.15.15/include --oldincludedir=/nix/store/gh4r6pp80ghgznzm57q096c3k9rckwq6-awscli2-2.15.15/include --mandir=/nix/store/gh4r6pp80ghgznzm57q096c3k9rckwq6-awscli2-2.15.15/share/man --infodir=/nix/store/gh4r6pp80ghgznzm57q096c3k9rckwq6-awscli2-2.15.15/share/info --docdir=/nix/store/gh4r6pp80ghgznzm57q096c3k9rckwq6-awscli2-2.15.15/share/doc/awscli --libdir=/nix/store/gh4r6pp80ghgznzm57q096c3k9rckwq6-awscli2-2.15.15/lib --libexecdir=/nix/store/gh4r6pp80ghgznzm57q096c3k9rckwq6-awscli2-2.15.15/libexec --localedir=/nix/store/gh4r6pp80ghgznzm57q096c3k9rckwq6-awscli2-2.15.15/share/locale
checking for a Python interpreter with version >= 3.8... python
checking for python... /nix/store/7wz6hm9i8wljz0hgwz1wqmn2zlbgavrq-python3-3.11.8/bin/python
checking for python version... 3.11
checking for python platform... linux
checking for GNU default python prefix... ${prefix}
checking for GNU default python exec_prefix... ${exec_prefix}
checking for python script directory (pythondir)... ${PYTHON_PREFIX}/lib/python3.11/site-packages
checking for python extension module directory (pyexecdir)... ${PYTHON_EXEC_PREFIX}/lib/python3.11/site-packages
checking for sqlite3... yes
checking for --with-install-type... system-sandbox
checking for --with-download-deps... Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/build/source/./backends/build_system/__main__.py", line 125, in <module>
    main()
  File "/build/source/./backends/build_system/__main__.py", line 121, in main
    parsed_args.func(parsed_args)
  File "/build/source/./backends/build_system/__main__.py", line 49, in validate
    validate_env(parsed_args.artifact)
  File "/build/source/./backends/build_system/validate_env.py", line 36, in validate_env
    unmet_deps = _get_unmet_dependencies(requirements)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/build/source/./backends/build_system/validate_env.py", line 81, in _get_unmet_dependencies
    if not requirement.is_in_range(actual_version):
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/build/source/./backends/build_system/utils.py", line 90, in is_in_range
    return self._meets_constraints(version)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/build/source/./backends/build_system/utils.py", line 93, in _meets_constraints
    return all(
           ^^^^
  File "/build/source/./backends/build_system/utils.py", line 94, in <genexpr>
    self._meets_constraint(version, constraint)
  File "/build/source/./backends/build_system/utils.py", line 103, in _meets_constraint
    version, constraint_version = self._normalize(version, constraint_version)
                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/build/source/./backends/build_system/utils.py", line 111, in _normalize
    v1_parts = [int(v) for v in v1.split(".")]
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/build/source/./backends/build_system/utils.py", line 111, in <listcomp>
    v1_parts = [int(v) for v in v1.split(".")]
                ^^^^^^
ValueError: invalid literal for int() with base 10: 'post0'
configure: error: "Python dependencies not met."
```

### Additional context

New to NixOS, I ran `nix flake update` and `sudo nixos-rebuild switch --upgrade --flake ~/nixos#pc`. Removing and reinstalling the package did not fix the issue.

### Notify maintainers

@bhipple @davegallant @bryanasdev000 @devusb @anthonyroussel 

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.22, NixOS, 24.05 (Uakari), 24.05.20240321.20f77aa`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.2`
 - channels(root): `"nixos-23.11"`
 - nixpkgs: `/nix/store/wa8lhchql76i6czzvad3bqz6lvnjnbr5-source`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/tools/admin/awscli2/default.nix
+++ b/pkgs/tools/admin/awscli2/default.nix
@@ -4,6 +4,7 @@
 , groff
 , less
 , fetchFromGitHub
+, fetchpatch
 , installShellFiles
 , nix-update-script
 , testers
@@ -17,6 +18,23 @@ let
         disabledTests = prev.disabledTests ++ [
           "test_check_link_response_only" # fails on hydra https://hydra.nixos.org/build/242624087/nixlog/1
         ];
+      });
+      python-dateutil = prev.python-dateutil.overridePythonAttrs (prev: {
+        version = "2.8.2";
+        pyproject = null;
+        src = prev.src.override {
+          version = "2.8.2";
+          hash = "sha256-ASPKzBYnrhnd88J6XeW9Z+5FhvvdZEDZdI+Ku0g9PoY=";
+        };
+        patches = [
+          # https://github.com/dateutil/dateutil/pull/1285
+          (fetchpatch {
+            url = "https://github.com/dateutil/dateutil/commit/f2293200747fb03d56c6c5997bfebeabe703576f.patch";
+            relative = "src";
+            hash = "sha256-BVEFGV/WGUz9H/8q+l62jnyN9VDnoSR71DdL+LIkb0o=";
+          })
+        ];
+        postPatch = null;
       });
       ruamel-yaml = prev.ruamel-yaml.overridePythonAttrs (prev: {
         src = prev.src.override {
@@ -41,27 +59,26 @@ let
 in
 with py.pkgs; buildPythonApplication rec {
   pname = "awscli2";
-  version = "2.15.15"; # N.B: if you change this, check if overrides are still up-to-date
+  version = "2.15.31"; # N.B: if you change this, check if overrides are still up-to-date
   pyproject = true;
 
   src = fetchFromGitHub {
     owner = "aws";
     repo = "aws-cli";
     rev = "refs/tags/${version}";
-    hash = "sha256-hJuJkCiKgSxfPVgS5II7BwpyQhjaRE2Ct3ZJQq6xWgg=";
+    hash = "sha256-7qeB3NCYQi7Mke/ZCGe8QUS5qy26btpGjhY/FMNhdkE=";
   };
 
   postPatch = ''
     substituteInPlace pyproject.toml \
-      --replace 'cryptography>=3.3.2,<40.0.2' 'cryptography>=3.3.2' \
-      --replace 'flit_core>=3.7.1,<3.8.1' 'flit_core>=3.7.1' \
-      --replace 'awscrt>=0.19.18,<=0.19.19' 'awscrt>=0.19.18' \
-      --replace 'docutils>=0.10,<0.20' 'docutils>=0.10' \
-      --replace 'prompt-toolkit>=3.0.24,<3.0.39' 'prompt-toolkit>=3.0.24'
+      --replace-fail 'awscrt>=0.19.18,<=0.19.19' 'awscrt>=0.19.18' \
+      --replace-fail 'cryptography>=3.3.2,<40.0.2' 'cryptography>=3.3.2' \
+      --replace-fail 'distro>=1.5.0,<1.9.0' 'distro>=1.5.0' \
+      --replace-fail 'docutils>=0.10,<0.20' 'docutils>=0.10' \
+      --replace-fail 'prompt-toolkit>=3.0.24,<3.0.39' 'prompt-toolkit>=3.0.24'
 
     substituteInPlace requirements-base.txt \
-      --replace "wheel==0.38.4" "wheel>=0.38.4" \
-      --replace "flit_core==3.8.0" "flit_core>=3.8.0"
+      --replace-fail "wheel==0.38.4" "wheel>=0.38.4"
 
     # Upstream needs pip to build and install dependencies and validates this
     # with a configure script, but we don't as we provide all of the packages
@@ -69,12 +86,12 @@ with py.pkgs; buildPythonApplication rec {
     sed -i '/pip>=/d' requirements/bootstrap.txt
   '';
 
-  nativeBuildInputs = [
+  build-system = [
     installShellFiles
     flit-core
   ];
 
-  propagatedBuildInputs = [
+  dependencies = [
     awscrt
     bcdoc
     botocore
@@ -116,6 +133,9 @@ with py.pkgs; buildPythonApplication rec {
   ];
 
   disabledTestPaths = [
+    "tests/dependencies"
+    "tests/unit/botocore"
+
     # Integration tests require networking
     "tests/integration"
 

